### PR TITLE
Support `conversation_style` of `openai` format (OpenAI API style)

### PR DESCRIPTION
**Pull request by xingyaoww**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

This adds a `conversation_style` of `openai`, similar to the `openai` one by @musabgultekin in https://github.com/pytorch/torchtune/issues/824#issuecomment-2068049678. 

#### Changelog
What are the changes made in this PR?

* Add conversion function of `openai_to_llama2_messages`

It now supports converting sample like (following [OpenAI API](https://platform.openai.com/docs/guides/text-generation/chat-completions-api)) the following directly to internal `llama2_messages`.

```
{
    # key could be "messages" OR "conversations"
    "messages": [
        {
            "role": <system|user|assistant>,
            "content": <message>,
        },
        ...
    ]
}
```

Example usage in config:
```
# Dataset
dataset:
  _component_: torchtune.datasets.chat_dataset
  source: json
  data_files: data/datasets/codeact-mixture.jsonl
  split: train
  conversation_style: standard_chat
  chat_format: ChatMLFormat
  max_seq_len: 16384
seed: 42
shuffle: True
```

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
	- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

### CODE REVIEW

File: torchtune/data/_converters.py [hunk 1]
```diff
@@ -55,3 +55,59 @@ def sharegpt_to_llama2_messages(
...
         messages.append(Message(role=role, content=content, masked=masked))
     return messages
+
+
+def standard_chat_to_llama2_messages(
+    sample: Mapping[str, Any],
+    train_on_input: bool = False,
+) -> List[Message]:
+    """
+    Convert a chat sample adhering to the OpenAI API standard chat format to the Llama2 chat format.
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:65]: Do you know if there's a reference for this?

File: torchtune/data/_converters.py [hunk 2]
```diff
@@ -55,3 +55,59 @@ def sharegpt_to_llama2_messages(
...
+            ...
+        ]
+
+    Args:
+        sample (Mapping[str, Any]): a single data sample with "conversations" field pointing
+            to a list of dict messages.
+        train_on_input (bool): whether the prompt should remain unmasked. Default: False
+        messages_key (str): the key in the sample that contains the messages. Default: "messages"
+
+    Returns:
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:96]: You should add a Raises: section here or doclint will complain

File: torchtune/data/_converters.py [hunk 3]
```diff
@@ -55,3 +55,59 @@ def sharegpt_to_llama2_messages(
...
+        messages_key = "messages"
+    elif "conversations" in sample:
+        messages_key = "conversations"
+    else:
+        raise ValueError(f"Sample does not contain 'messages' or 'conversations' key. Existing keys: {sample.keys()}")
+    conversations = sample[messages_key]
+
+    messages = []
+    for message in conversations:
+        role = message["role"]
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:109]: nit: you could also use the `Message.from_dict()` method. It won't save you many lines of code, hence the nit :) this was just recently added, which is why the sharegpt converter doesn't use it yet

File: torchtune/data/_converters.py [hunk 4]
```diff
@@ -55,3 +55,59 @@ def sharegpt_to_llama2_messages(
         masked = (role != "assistant") and (not train_on_input)
         messages.append(Message(role=role, content=content, masked=masked))
     return messages
+
+
+def standard_chat_to_llama2_messages(
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:60]: Why call this standard chat, and not openai_to_llama2_messages?

**Review by RdoubleA — COMMENTED** [review]
Thanks @xingyaoww for this addition! Overall looks great, just some minor comments.

Would be great if you can add a unit test for this as well. I also realized the sharegpt converter is missing a unit test also (I know, how could we) but both of these should be tested. Upto you if you also want to cover that :)

File: torchtune/data/_converters.py (hunk 1 above)
Comment (xingyaoww) [reply torchtune/data/_converters.py:65]: Yes! We do at https://platform.openai.com/docs/guides/text-generation/chat-completions-api

I will add that to the code :)

File: torchtune/data/_converters.py (hunk 4 above)
Comment (xingyaoww) [reply torchtune/data/_converters.py:60]: Will rename! I just came up with this random name for some reason..

File: tests/torchtune/data/test_converters.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            "from": "assistant",
+            "value": CHAT_SAMPLE["assistant"],
+        },
+    ]
+
+    def test_conversion(self):
+        converted_messages = sharegpt_to_llama2_messages(self.samples)
+        for converted, expected in zip(converted_messages, EXPECTED_MESSAGE):
+            assert converted == expected
```
Comment (RdoubleA) [inline tests/torchtune/data/test_converters.py:53]: I'm not sure if equality between `Message` classes will work here, you can either check each field directly, add a `to_dict` method to `Message` and compare the dictionaries, or implement an `__eq__` method for `Message`. I think the `to_dict` method is probably easiest.

File: tests/torchtune/data/test_converters.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    def test_conversion_train_on_input(self):
+        converted_messages = sharegpt_to_llama2_messages(
+            self.samples, train_on_input=True
+        )
+        for converted, expected in zip(converted_messages, EXPECTED_MESSAGE):
+            assert converted == expected
```
Comment (RdoubleA) [inline tests/torchtune/data/test_converters.py:60]: do you also want to check that the mask is set properly? by default the mask is false, so you might want to set `train_on_input=False` to set mask on user message to True and properly test this functionality.

File: tests/torchtune/data/test_converters.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        content=CHAT_SAMPLE["user"],
+    ),
+    Message(
+        role="assistant",
+        content=CHAT_SAMPLE["assistant"],
+    ),
+]
+
+
+class TestSharegptToLlama2Messages:
```
Comment (RdoubleA) [inline tests/torchtune/data/test_converters.py:34]: nit:
```suggestion
class TestShareGPTToLlama2Messages:
```

File: tests/torchtune/data/test_converters.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+    def test_conversion_train_on_input(self):
+        converted_messages = sharegpt_to_llama2_messages(
+            self.samples, train_on_input=True
+        )
+        for converted, expected in zip(converted_messages, EXPECTED_MESSAGE):
+            assert converted == expected
+
+
+class TestOpenaiToLlama2Messages:
```
Comment (RdoubleA) [inline tests/torchtune/data/test_converters.py:63]: nit:
```suggestion
class TestOpenAIToLlama2Messages:
```

File: tests/torchtune/data/test_converters.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    samples_1 = {
+        "id": "DUMMY",
+        "conversations": [
+            {
+                "from": "system",
```
Comment (RdoubleA) [inline tests/torchtune/data/test_converters.py:68]: I thought the OpenAI format doesn't use from / value fields?

File: torchtune/data/_converters.py [hunk 10]
```diff
@@ -55,3 +55,61 @@ def sharegpt_to_llama2_messages(
...
+
+
+def openai_to_llama2_messages(
+    sample: Mapping[str, Any],
+    train_on_input: bool = False,
+) -> List[Message]:
+    """
+    Convert a chat sample adhering to the OpenAI API standard chat format to the Llama2 chat format.
+
+    OpenAI API standard chat format follows::
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:67]: you can hyperlink the text like this so that the live docs parse it correctly:
```suggestion
    OpenAI API `standard chat format <https://platform.openai.com/docs/guides/text-generation/chat-completions-api>`_ follows::
```

File: tests/torchtune/data/test_converters.py (hunk 5 above)
Comment (xingyaoww) [reply tests/torchtune/data/test_converters.py:53]: yes there's a bunch of issue with this test - I'll fix it in the next commit but there's something wrong with my local pre-commit that I'm still fixing :(

File: tests/torchtune/data/test_converters.py (hunk 6 above)
Comment (xingyaoww) [reply tests/torchtune/data/test_converters.py:60]: yes! in the next commit!

File: tests/torchtune/data/test_converters.py (hunk 9 above)
Comment (xingyaoww) [reply tests/torchtune/data/test_converters.py:68]: typo - will fix in the next commit

**Comment by xingyaoww** [comment]
@RdoubleA Thanks a ton for the review!!! I should fixed all the test issues in the latest commit and can confirm all tests passed!

image

File: torchtune/data/_converters.py [hunk 11]
```diff
@@ -55,3 +55,61 @@ def sharegpt_to_llama2_messages(
...
+
+def openai_to_llama2_messages(
+    sample: Mapping[str, Any],
+    train_on_input: bool = False,
+) -> List[Message]:
+    """
+    Convert a chat sample adhering to the OpenAI API standard chat format to the Llama2 chat format.
+
+    OpenAI API `standard chat format <https://platform.openai.com/docs/guides/text-generation/chat-completions-api>`_ follows::
+    Check https://platform.openai.com/docs/guides/text-generation/chat-completions-api for more details
```
Comment (RdoubleA) [inline torchtune/data/_converters.py:68]: remove this line?

**Review by RdoubleA — APPROVED** [review]
this looks good to go! and thanks for the additional test! let me start the github CI and see what it thinks

File: torchtune/data/_converters.py (hunk 11 above)
Comment (xingyaoww) [reply torchtune/data/_converters.py]: done!

**Comment by RdoubleA** [comment]
@xingyaoww Looks like just a lint error which you should be able to fix with `pre-commit`

**Comment by xingyaoww** [comment]
@RdoubleA should be fixed!
